crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2673,6 +2722,73 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    /// Performs lexicographic comparison between two slices of types.
+    ///
+    /// This comparison supports only equality and ordering operators: `==`, `!=`, `<`, `<=`, `>`, `>=`.
+    /// For lexicographic comparison, elements from both slices are compared pairwise using
+    /// `infer_binary_type_comparison`. If a conclusive result cannot be determined as a `BoolLiteral`,
+    /// it returns `bool`. If the comparison is undefined between types, it returns `None`.
+    fn infer_lexicographic_type_comparison(
+        &mut self,
+        left: &[Type<'db>],
+        op: ast::CmpOp,
+        right: &[Type<'db>],
+    ) -> Option<Type<'db>> {
+        if !matches!(
+            op,
+            ast::CmpOp::Eq
+                | ast::CmpOp::NotEq
+                | ast::CmpOp::Lt
+                | ast::CmpOp::LtE
+                | ast::CmpOp::Gt
+                | ast::CmpOp::GtE
+        ) {
+            return None;
+        }
+
+        // Compare paired elements from left and right slices
+        for (l_ty, r_ty) in left.iter().zip(right.iter()) {
+            let eq_result_ty = self.infer_binary_type_comparison(*l_ty, ast::CmpOp::Eq, *r_ty);
+
+            match eq_result_ty {
+                // Types are equal, continue to the next pair
+                Some(Type::BooleanLiteral(true)) => continue,
+
+                // Types are not equal, perform the specified comparison and return the result
+                Some(Type::BooleanLiteral(false)) => {
+                    return self.infer_binary_type_comparison(*l_ty, op, *r_ty)
+                }
+
+                // If the intermediate result is a bool instance, we cannot determine the final result using BoolLiteral.
+                // In this case, we simply return a bool instance.
+                Some(Type::Instance(class_ty)) if class_ty.is_known(self.db, KnownClass::Bool) => {
+                    return Some(KnownClass::Bool.to_instance(self.db));
+                }
+
+                // TODO: Todo propagation; it should be removed later.
+                Some(Type::Todo) => return Some(Type::Todo),
+
+                // If comparison is undefined, return None
+                _ => return None,
+            }
```

Comment:
It is possible for an equality comparison to be supported but return a non-bool type. See the runtime behavior:

```
>>> class A:
...     def __eq__(self, other):
...
>>> (1, A()) == (1, 4)
True
```

Note the tuple comparison swallows the non-bool type from `A.__eq__` and ends up always returning `True` or `False`.

So I think our handling here can/should be a bit more general, we can just call `Type::bool` on any returned type and consider it equal / not equal / unknown according to that `Truthiness` result? The only special case I think we might want is the Type::Todo propagation.

And similar to the comments above, I think `Eq` comparison should never return `None`, so there also shouldn't be any case here where we return `None` explicitly; though we might pass through a `None` from the call back to `infer_binary_type_comparison` with a non-Eq op.